python-checks/src/main/java/org/sonar/python/checks/hotspots/GraphQLIntrospectionCheck.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks.hotspots;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.ExpressionList;
+import org.sonar.plugins.python.api.tree.ListLiteral;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tuple;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6786")
+public class GraphQLIntrospectionCheck extends PythonSubscriptionCheck {
+
+  private static final Set<String> GRAPHQL_VIEWS_FQNS = Set.of(
+    "flask_graphql.GraphQLView.as_view",
+    "graphql_server.flask.GraphQLView.as_view");
+
+  private static final Set<String> SENSITIVE_VALIDATION_RULE_FQNS = Set.of(
+    "graphene.validation.DisableIntrospection",
+    "graphql.validation.NoSchemaIntrospectionCustomRule");
+
+  private static final String MESSAGE = "Disable introspection on this GraphQL server endpoint.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, GraphQLIntrospectionCheck::checkGraphQLIntrospection);
+  }
+
+  private static void checkGraphQLIntrospection(SubscriptionContext ctx) {
+    CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+    Optional.ofNullable(callExpression.calleeSymbol())
+      .map(Symbol::fullyQualifiedName)
+      .filter(GRAPHQL_VIEWS_FQNS::contains)
+      .filter(fqn -> hasSensitiveMiddlewares(callExpression.arguments()) ||
+        hasSensitiveValidationRules(callExpression.arguments()))
```

Comment:
I think that the logical operator here should be `&&` rather than `||`. 
If either a middleware or a validation rule disables introspection, we should not raise an issue.

(Admittedly, the ticket says `When the .as_view(...) method is called on these classes, S6786 should be raised if there is no middleware or no validator that disables introspection.`, but looking at the RSPEC, it seems that the issue should be raised only if both are missing).

Note: Maybe instead of naming the methods `hasSensitiveMiddlewares`, we could consider having them named `doesNotHaveSafeMiddleware` and `doesNotHaveSafeValidationRule` to reflect more explicitly what the rule is looking for.